File: src/main/kotlin/web/client/utils/HostCallsQueue.kt
package web.client.utils

import io.github.oshai.kotlinlogging.KotlinLogging
import time.UTC
import web.url.Url
import java.util.concurrent.ConcurrentHashMap

private const val DELAY_FOR_HOST_SEC = 5

private val logger = KotlinLogging.logger {}

class HostCallsQueue {
    /**
     * Key = hostname. Value = Next allowed call timestamp (epoch sec).
     */
    private val hostnameToTime = ConcurrentHashMap<String, Long>()

    fun <T> patiently(url: Url, call: () -> T): T {
        waitUtilCallAllowed(url)

        val result = call()

        hostnameToTime[url.getHost()] = UTC.Now.epochSec() + DELAY_FOR_HOST_SEC

        return result
    }

    private fun waitUtilCallAllowed(url: Url) {
        val waitUntilEpochSec = hostnameToTime.computeIfAbsent(url.getHost()) { UTC.Now.epochSec() }

        val secondsToWait = waitUntilEpochSec - UTC.Now.epochSec()

        if (secondsToWait > 0) {
            logger.info("Pausing for $secondsToWait seconds for ${url.getUrl()}")

            Thread.sleep(secondsToWait * 1000)
        }
    }
}

File: src/Utils/Web/HostCallsQueue.php
<?php

declare(strict_types=1);

namespace App\Utils\Web;

use App\Utils\DateTime\UtcClock;
use App\Utils\Web\Url\Url;
use Veelkoov\Debris\Maps\StringToInt;

class HostCallsQueue
{
    /**
     * Key = hostname. Value = Next allowed call timestamp (epoch sec).
     */
    private readonly StringToInt $hostToTimestamp;

    public function __construct(
        private readonly int $delayForHostSec,
    ) {
        $this->hostToTimestamp = new StringToInt();
    }

    /**
     * @template T
     *
     * @param callable(): T $callback
     *
     * @return T
     */
    public function patiently(Url $url, callable $callback): mixed
    {
        $host = UrlUtils::getHost($url->getUrl());

        $this->waitUtilCallAllowed($host);
        $result = $callback();
        $this->hostToTimestamp->set($host, UtcClock::time() + $this->delayForHostSec);

        return $result;
    }

    private function waitUtilCallAllowed(string $host): void
    {
        $waitUntilEpochSec = $this->hostToTimestamp->getOrDefault($host, static fn () => UtcClock::time());

        $secondsToWait = $waitUntilEpochSec - UtcClock::time();

        if ($secondsToWait > 0) {
            UtcClock::sleep($secondsToWait);
        }
    }
}
